drivers/flash/flash_ambiq.c

```diff
@@ -0,0 +1,192 @@
+#define FLASH_ERASE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+BUILD_ASSERT((FLASH_WRITE_BLOCK_SIZE & (MIN_WRITE_SIZE - 1)) == 0,
+	     "The flash write block size must be a multiple of 16!");
+
+#define FLASH_ERASE_BYTE 0xFF
+#define FLASH_ERASE_WORD                                                                           \
+	(((uint32_t)(FLASH_ERASE_BYTE << 24)) | ((uint32_t)(FLASH_ERASE_BYTE << 16)) |             \
+	 ((uint32_t)(FLASH_ERASE_BYTE << 8)) | ((uint32_t)FLASH_ERASE_BYTE))
+
+#if defined(CONFIG_MULTITHREADING)
+static struct k_sem flash_ambiq_sem;
+#define FLASH_SEM_INIT() k_sem_init(&flash_ambiq_sem, 1, 1)
+#define FLASH_SEM_TAKE() k_sem_take(&flash_ambiq_sem, K_FOREVER)
+#define FLASH_SEM_GIVE() k_sem_give(&flash_ambiq_sem)
+#else
+#define FLASH_SEM_INIT()
+#define FLASH_SEM_TAKE()
+#define FLASH_SEM_GIVE()
+#endif /* CONFIG_MULTITHREADING */
+
+static const struct flash_parameters flash_ambiq_parameters = {
+	.write_block_size = FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = FLASH_ERASE_BYTE,
+};
+
+static bool flash_ambiq_valid_range(off_t offset, size_t len)
+{
+	if ((offset < 0) || offset >= SOC_NV_FLASH_SIZE || (SOC_NV_FLASH_SIZE - offset) < len) {
+		return false;
+	}
+
+	return true;
+}
+
+static int flash_ambiq_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	memcpy(data, (uint8_t *)(SOC_NV_FLASH_ADDR + offset), len);
+
+	return 0;
+}
+
+static int flash_ambiq_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	int ret = 0;
+	uint32_t critical = 0;
+	uint32_t aligned[FLASH_WRITE_BLOCK_SIZE / sizeof(uint32_t)] = {0};
+	uint32_t *src = (uint32_t *)data;
+
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* write address must be block size aligned and the write length must be multiple of block
+	 * size.
+	 */
+	if (((uint32_t)offset & (FLASH_WRITE_BLOCK_SIZE - 1)) ||
+	    (len & (FLASH_WRITE_BLOCK_SIZE - 1))) {
+		return false;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	FLASH_SEM_TAKE();
+
+	critical = am_hal_interrupt_master_disable();
+	for (int i = 0; i < len / FLASH_WRITE_BLOCK_SIZE; i++) {
+		for (int j = 0; j < FLASH_WRITE_BLOCK_SIZE / sizeof(uint32_t); j++) {
+			/* Make sure the source data is 4-byte aligned. */
+			aligned[j] = UNALIGNED_GET((uint32_t *)src);
+			src++;
+		}
+		ret = am_hal_mram_main_program(
+			AM_HAL_MRAM_PROGRAM_KEY, aligned,
+			(uint32_t *)(SOC_NV_FLASH_ADDR + offset + i * FLASH_WRITE_BLOCK_SIZE),
+			FLASH_WRITE_BLOCK_SIZE / sizeof(uint32_t));
+		if (ret) {
+			break;
+		}
+	}
+	am_hal_interrupt_master_set(critical);
+
+	FLASH_SEM_GIVE();
+
+	return ret;
+}
+
+static int flash_ambiq_erase(const struct device *dev, off_t offset, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	int ret = 0;
+
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
```

Comment:
The alignment check should also be added here.

Reply:
@Laczen I added alignment check here in the previous implementation, but @de-nordic commented we should put it to flash_write API since only flash_write needs alignment check, so I modified it. How do you think about it?

Reply:
For erase operation, there is address alignment and length check in the HAL API, so I don't think we need to add it. I can leave a comment in the code.

Reply:
Yes, in that case it is not needed. As this is one hal operation we can rely on the check by the hal (I forgot about this). You can disregard this remark.

A little comment line would be welcome for future users to understand why a check is not needed.